include/swift/IDE/CodeCompletion.h

```diff
@@ -612,214 +617,357 @@ enum class CompletionKind {
-        TypeDistance(unsigned(TypeDistance)),
-        DiagnosticSeverity(unsigned(diagSeverity)),
-        DiagnosticMessage(DiagnosticMessage) {
-    AssociatedKind = static_cast<unsigned>(DeclKind);
-    assert(CompletionString);
-    assert(!isOperator() ||
-           getOperatorKind() != CodeCompletionOperatorKind::None);
+  CodeCompletionResultKind getKind() const { return Kind; }
+
+  /// Returns the raw associated kind which will be interpreted differently
+  /// depending on the completion item's kind.
+  uint8_t getRawAssociatedKind() const { return AssociatedKind; }
+
+  CodeCompletionDeclKind getAssociatedDeclKind() const {
+    assert(getKind() == CodeCompletionResultKind::Declaration);
+    return static_cast<CodeCompletionDeclKind>(AssociatedKind);
+  }
+  CodeCompletionLiteralKind getLiteralKind() const {
+    assert(getKind() == CodeCompletionResultKind::Literal);
+    return static_cast<CodeCompletionLiteralKind>(AssociatedKind);
+  }
+  CodeCompletionKeywordKind getKeywordKind() const {
+    assert(getKind() == CodeCompletionResultKind::Keyword);
+    return static_cast<CodeCompletionKeywordKind>(AssociatedKind);
+  }
+
+  CodeCompletionOperatorKind getKnownOperatorKind() const {
+    assert(isOperator());
+    return KnownOperatorKind;
+  }
+
+  bool isSystem() const { return IsSystem; };
+
+  CodeCompletionString *getCompletionString() const { return CompletionString; }
+
+  StringRef getModuleName() const { return ModuleName; }
+
+  StringRef getBriefDocComment() const { return BriefDocComment; }
+
+  ArrayRef<StringRef> getAssociatedUSRs() const { return AssociatedUSRs; }
+
+  const CodeCompletionResultType &getResultType() const { return ResultType; }
+
+  ContextFreeNotRecommendedReason getNotRecommendedReason() const {
+    return NotRecommended;
+  }
+
+  CodeCompletionDiagnosticSeverity getDiagnosticSeverity() const {
+    return DiagnosticSeverity;
+  }
+  StringRef getDiagnosticMessage() const { return DiagnosticMessage; };
+
+  bool isOperator() const {
+    if (getKind() == CodeCompletionResultKind::Declaration) {
+      switch (getAssociatedDeclKind()) {
+      case CodeCompletionDeclKind::PrefixOperatorFunction:
+      case CodeCompletionDeclKind::PostfixOperatorFunction:
+      case CodeCompletionDeclKind::InfixOperatorFunction:
+        return true;
+      default:
+        return false;
+      }
+    } else {
+      return getKind() == CodeCompletionResultKind::BuiltinOperator;
+    }
+  }
+
+  /// Helper methods used during initialization of
+  /// \c ContextFreeCodeCompletionResult.
+  static CodeCompletionOperatorKind
+  getCodeCompletionOperatorKind(CodeCompletionString *str);
+  static CodeCompletionDeclKind getCodeCompletionDeclKind(const Decl *D);
+  static bool getDeclIsSystem(const Decl *D);
+};
+
+/// A single code completion result enriched with information that depend on
+/// the completion's usage context.
+class CodeCompletionResult {
+  ContextFreeCodeCompletionResult ContextFree;
```

Comment:
`ContextFreeCodeCompletionResult` is fairly big object. Instead of copying it, could `CodeCompletionResult` just hold a reference to it?

Reply:
I thought about the same but decided against it, because 
- this way, we are sure that `ContextFreeCodeCompletionResult` is allocated by the same allocator as `CodeCompletionResult` and we don’t run into issues that `ContextFreeCodeCompletionResult` might get deallocated while `CodeCompletionResult` is still referencing it
- we only need to copy it when constructing `CodeCompletionResult` and I don’t think it makes much of a performance difference
- We don’t have to do an indirection to `ContextFreeCodeCompletionResult` when accessing context-free properties of a `CodeCompletionResult`

Do you think this makes sense or would you think it’s worth to measure the performance impact of the copy?